coc/ext/fullwarapi/__init__.py

```diff
@@ -0,0 +1,221 @@
+        return None
+
+    try:
+        signing, _ = token.rsplit(b".", 1)
+        _, payload = signing.split(b".", 1)
+    except ValueError:
+        return None  # not enough segments
+
+    if len(payload) % 4 > 0:
+        payload += b"=" * (4 - len(payload) % 4)
+
+    bytes_payload = base64.urlsafe_b64decode(payload)
+    dict_payload = json.loads(bytes_payload)
+    try:
+        expiry = dict_payload["exp"]
+        return datetime.fromtimestamp(expiry)
+    except KeyError:
+        return None
+
+
+async def login(username: str, password: str, clash_client) -> "FullWarClient":
+    """Eases logging into the API client.
+
+    For more information on this project, please join the discord server - <discord.gg/Eaja7gJ>
+
+    You must have your username and password as given on the server.
+    If unsure as to what this means, please reach out to an admin.
+
+    Parameters
+    -----------
+    username : str
+        Your username as given on the discord server.
+    password : str
+        Your password as given on the discord server
+    loop : Optional[:class:`asyncio.AbstractEventLoop`]
+        The :class:`asyncio.AbstractEventLoop` to use for HTTP requests.
+        An :func:`asyncio.get_event_loop()` will be used if ``None`` is passed
+    """
+    if not isinstance(username, str) or not isinstance(password, str):
+        raise TypeError("username and password must both be a string")
+    if not username or not password:
+        raise ValueError("username or password must not be an empty string.")
+
+    loop = asyncio.get_running_loop()
+    return FullWarClient(username, password, clash_client, loop)
+
+
+class FullWarClient:
+    """An extension that helps interact with the Full War API.
+
+    For more information on this project, please join the discord server - <discord.gg/Eaja7gJ>
+
+    You must have your username and password as given on the server.
+    If unsure as to what this means, please reach out to an admin.
+
+    Parameters
+    -----------
+    username : str
+        Your username as given on the discord server.
+    password : str
+        Your password as given on the discord server
+    clash_client: coc.Client
+        Client to use
+
+    loop : Optional[:class:`asyncio.AbstractEventLoop`]
+        The :class:`asyncio.AbstractEventLoop` to use for HTTP requests.
+        An :func:`asyncio.get_event_loop()` will be used if ``None`` is passed
+
+    """
+
+    BASE_URL = "https://fw-api.teamutils.com"
+
+    __slots__ = ("username", "password", "clash_client", "loop", "key", "http_session")
+
+    def __init__(self, username: str, password: str, clash_client, loop: asyncio.AbstractEventLoop = None):
+        self.username = username
+        self.password = password
+        self.clash_client = clash_client
+
+        self.loop = loop or asyncio.get_event_loop()
+        self.key = None  # set in get_key()
+
+        self.http_session = aiohttp.ClientSession(loop=self.loop)
```

Comment:
Similar comment to above, see https://docs.aiohttp.org/en/stable/client_reference.html, this parameter is deprecated, is there a reason why we're passing a loop in?